Add node serial e2e tests that simulate the kubelet restart

gjkim42 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR adds node e2e tests to make sure a completed init container is not restarted due to the kubelet restart.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #
part of the https://github.com/kubernetes/kubernetes/issues/119014

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
none
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

gjkim42 commented:
/test

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-containerd

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-containerd

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2

aojea commented on test/e2e_node/container_lifecycle_test.go:
```diff
@@ -596,3 +598,320 @@ var _ = SIGDescribe("[NodeConformance] Containers Lifecycle ", func() {
+			startKubelet()
+			// wait until the kubelet health check will succeed
+			gomega.Eventually(ctx, func() bool {
+				return kubeletHealthCheck(kubeletHealthCheckURL)
```
cdan this race with the restart? i.e., it succeeds because kubelet didn't already restart?
it will be ideal if we have some kind of uptime metric on the kubelet to ensure this is really a new started kubelet

gjkim42 replied:
To prevent that, wait for the health check to be fail before restarting it.

aojea commented on test/e2e_node/container_lifecycle_test.go:
```diff
@@ -596,3 +598,320 @@ var _ = SIGDescribe("[NodeConformance] Containers Lifecycle ", func() {
+			}, f.Timeouts.PodStart, f.Timeouts.Poll).Should(gomega.BeTrue())
+
+			ginkgo.By("waiting for the kubelet to sync the pod")
+			time.Sleep(1 * time.Minute)
```
can we wait based on some condition, sleeps like these cause a lot of flakes

gjkim42 replied:
Actually, we expect the kubelet do nothing here. Not sure which condition is deterministic...

aojea replied:
> waiting for the kubelet to sync the pod

you should have some way to know that the pod is synced, shouldn't you?

gjkim42 replied:
I think we can stop the regular container as well before restarting the kubelet, and check if the container is restarted which means the pod is synced by the kubelet.

I'll fix it.

gjkim42 replied:
I thought about this for a while and I think this won't cause any flake.

This spec aims to ensure that the kubelet is consistently not making any action for the init container, so it may rather cause a false positive result, but if the test failed it should be the real failure.

What do you think?

gjkim42 replied:
updated the test spec to ensure that any completed init container does not restart consistently.

aojea commented on test/e2e_node/container_lifecycle_test.go:
```diff
@@ -596,3 +598,320 @@ var _ = SIGDescribe("[NodeConformance] Containers Lifecycle ", func() {
+			}, f.Timeouts.PodStart, f.Timeouts.Poll).Should(gomega.BeTrue())
+
+			ginkgo.By("waiting for the kubelet to sync the pod")
+			time.Sleep(1 * time.Minute)
```
ditto https://github.com/kubernetes/kubernetes/pull/119019/files#r1249437538

aojea commented on test/e2e_node/container_lifecycle_test.go:
```diff
@@ -596,3 +598,320 @@ var _ = SIGDescribe("[NodeConformance] Containers Lifecycle ", func() {
+			}, f.Timeouts.PodStart, f.Timeouts.Poll).Should(gomega.BeTrue())
+
+			ginkgo.By("waiting for the kubelet to sync the pod")
+			time.Sleep(1 * time.Minute)
```
https://github.com/kubernetes/kubernetes/pull/119019/files#r1249437538

aojea commented on test/e2e_node/container_lifecycle_test.go:
```diff
@@ -596,3 +598,320 @@ var _ = SIGDescribe("[NodeConformance] Containers Lifecycle ", func() {
+			}, f.Timeouts.PodStart, f.Timeouts.Poll).Should(gomega.BeTrue())
+
+			ginkgo.By("waiting for the kubelet to sync the pod")
+			time.Sleep(1 * time.Minute)
```
https://github.com/kubernetes/kubernetes/pull/119019/files#r1249437538

gjkim42 commented:
/cc @SergeyKanzhelev @smarterclayton

ffromani commented:
/triage accepted
/priority important-soon

I think we have already some restart-related tests but not specific in this area. Seems like a good addition

gjkim42 commented:
/retest

gjkim42 commented:
/test pull-kubernetes-node-e2e-containerd-sidecar-containers
/test pull-kubernetes-e2e-gce-cos-alpha-features

gjkim42 commented:
/test pull-kubernetes-e2e-gce-cos-alpha-features

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-containerd

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-containerd

ffromani commented:
/cc

gjkim42 committed:
Add node serial e2e tests that simulate the kubelet restart

This adds node e2e tests to make sure a completed init container is not
restarted due to the kubelet restart.

gjkim42 commented:
/test pull-kubernetes-node-e2e-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers

bart0sh commented:
/retest
/lifecycle-remove stale

bart0sh commented:
/remove-lifecycle stale
/cc @bart0sh @kannon92

bart0sh commented:
@gjkim42 please, fix failing CI tests, thanks.

gjkim42 commented:
/test pull-kubernetes-node-e2e-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers

gjkim42 commented:
/retest

gjkim42 commented:
/test

gjkim42 commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial

gjkim42 commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial

swatisehgal commented:
/test pull-kubernetes-node-kubelet-serial-containerd-alpha-features

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers

bart0sh commented on test/e2e_node/container_lifecycle_test.go:
```diff
@@ -886,6 +887,370 @@ var _ = SIGDescribe(framework.WithSerial(), "Containers Lifecycle", func() {
+			// wait until the kubelet health check will succeed
+			gomega.Eventually(ctx, func() bool {
+				return kubeletHealthCheck(kubeletHealthCheckURL)
+			}, f.Timeouts.PodStart, f.Timeouts.Poll).Should(gomega.BeTrueBecause("kubelet should be started"))
```
Would it make sense to enhance [restartKubelet](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/util.go#L436) (if needed) and use it here?

gjkim42 replied:
That may make sense, but I think the current pattern (stopKubelet + do whatever we need + startKubelet do whatever we need) is convenient enough.

bart0sh commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-kubelet-serial-containerd

SergeyKanzhelev commented:
@gjkim42 curious if you plan to resurrect this PR?

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-kubelet-serial-containerd

gjkim42 commented:
> @gjkim42 curious if you plan to resurrect this PR?

I'll take a look this week.

gjkim42 commented:
I guess the PR is ready to be reviewed. The only problem might be the test infra.

gjkim42 commented:
It seems there is a problem in my tests, I'll take a look soon.

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-kubelet-serial-containerd

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-kubelet-serial-containerd

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-containerd

gjkim42 commented:
ready to be reviewed.
/cc @SergeyKanzhelev

gjkim42 commented:
/open

gjkim42 commented:
/reopen

gjkim42 commented:
/remove-lifecycle rotten

toVersus commented on test/e2e_node/container_lifecycle_pod_construction.go:
```diff
@@ -336,14 +336,15 @@ func parseOutput(ctx context.Context, f *framework.Framework, pod *v1.Pod) conta
 		}
 		timestamp, err := strconv.ParseInt(fields[0], 10, 64)
-		framework.ExpectNoError(err)
+		framework.ExpectNoError(err, "Failed to parse the log: %q", log)
```
```suggestion
		framework.ExpectNoError(err, "Failed to parse the log timestamp: %q", log)
```

toVersus commented on test/e2e_node/container_lifecycle_pod_construction.go:
```diff
@@ -336,14 +336,15 @@ func parseOutput(ctx context.Context, f *framework.Framework, pod *v1.Pod) conta
+		framework.ExpectNoError(err, "Failed to parse the log: %q", log)
 		timeSinceBoot, err := strconv.ParseFloat(fields[1], 64)
-		framework.ExpectNoError(err)
+		framework.ExpectNoError(err, "Failed to parse the log: %q", log)
```
```suggestion
		framework.ExpectNoError(err, "Failed to parse the time since the host booted from the log: %q", log)
```

toVersus commented on test/e2e_node/container_lifecycle_test.go:
```diff
@@ -886,6 +887,355 @@ var _ = SIGDescribe(framework.WithSerial(), "Containers Lifecycle", func() {
+			}, 1*time.Minute, f.Timeouts.Poll).Should(gomega.BeTrueBecause("no completed init container should be restarted"))
+
+			ginkgo.By("Analyzing results")
+			// Cannot analyze the results with the container logs as the
```
Maybe we should test the behavior mentioned in the comment. For example, check if there are no startup logs for initContainers removed from the container runtime, while the startup logs for regular containers are still there. Right now, if the container status can't be found, it looks like the logs we get have an error message saying `"unable to retrieve container logs for <containerID>"`. This messes up the log parsing.

I'm not sure if this is the right way to go, but how about changing it so that the error message doesn't get written to the logs when the container status can't be found? You can see a similar handing [here](https://github.com/kubernetes/kubernetes/blob/v1.30.2/test/e2e/storage/podlogs/podlogs.go#L219) as well. So, something like this:

```diff
diff --git a/test/e2e/framework/pod/resource.go b/test/e2e/framework/pod/resource.go
index 3039a9bc821..ca96aa374cf 100644
--- a/test/e2e/framework/pod/resource.go
+++ b/test/e2e/framework/pod/resource.go
@@ -474,6 +474,9 @@ func getPodLogsInternal(ctx context.Context, c clientset.Interface, namespace, p
        if strings.Contains(string(logs), "Internal Error") {
                return "", fmt.Errorf("Fetched log contains \"Internal Error\": %q", string(logs))
        }
+       if strings.Contains(string(logs), "unable to retrieve container logs for ") {
+               return "", fmt.Errorf("Missing container status for %q: %q", containerName, string(logs))
+       }
        return string(logs), err
 }
```

Then, in the test case, we'll verify that only the regular container is starting.

```diff
diff --git a/test/e2e_node/container_lifecycle_test.go b/test/e2e_node/container_lifecycle_test.go
index 8dcb758bbab..895f49e3fa0 100644
--- a/test/e2e_node/container_lifecycle_test.go
+++ b/test/e2e_node/container_lifecycle_test.go
@@ -1051,9 +1051,19 @@ var _ = SIGDescribe(framework.WithSerial(), "Containers Lifec
ycle", func() {
                                return true
                        }, 1*time.Minute, f.Timeouts.Poll).Should(gomega.BeTrueBecau
se("no completed init container should be restarted"))

+                       ginkgo.By("Parsing results")
+                       pod, err = client.Get(ctx, pod.Name, metav1.GetOptions{})
+                       framework.ExpectNoError(err)
+                       results := parseOutput(ctx, f, pod)
+
                        ginkgo.By("Analyzing results")
-                       // Cannot analyze the results with the container logs as the
-                       // container statuses have been removed from container runtime.
+                       ginkgo.By("Verifying that only the regular container starts because the statues of init containers have been removed from container runtime")
+                       framework.ExpectNoError(results.DoesntStart(init1))
+                       framework.ExpectNoError(results.DoesntStart(init2))
+                       framework.ExpectNoError(results.DoesntStart(init3))
+                       framework.ExpectNoError(results.Starts(regular1))
+
+                       ginkgo.By("Verifying that the init containers have not been restarted")
                        gomega.Expect(pod.Status.InitContainerStatuses[0].RestartCount).To(gomega.Equal(int32(0)))
                        gomega.Expect(pod.Status.InitContainerStatuses[1].RestartCount).To(gomega.Equal(int32(0)))
                        gomega.Expect(pod.Status.InitContainerStatuses[2].RestartCount).To(gomega.Equal(int32(0)))
```

gjkim42 replied:
I am not sure checking the unstructured error message here is a good idea. We will see how other reviewers' comments.

Anyway, fixed it to check the regular1 is running with podStatus, thanks.

toVersus commented on test/e2e_node/container_lifecycle_test.go:
```diff
@@ -886,6 +887,355 @@ var _ = SIGDescribe(framework.WithSerial(), "Containers Lifecycle", func() {
+			client = e2epod.NewPodClient(f)
+			pod = client.Create(ctx, pod)
+			ginkgo.By("Waiting for the pod to be initializing the long-running init container")
+			err := e2epod.WaitForPodCondition(ctx, f.ClientSet, pod.Namespace, pod.Name, "PodInitializing", 1*time.Minute, func(pod *v1.Pod) (bool, error) {
```
```suggestion
			err := e2epod.WaitForPodCondition(ctx, f.ClientSet, pod.Namespace, pod.Name, "long-running init container initializing", 1*time.Minute, func(pod *v1.Pod) (bool, error) {
```

toVersus commented on test/e2e_node/container_lifecycle_test.go:
```diff
@@ -886,6 +887,355 @@ var _ = SIGDescribe(framework.WithSerial(), "Containers Lifecycle", func() {
+					if c.Name != longRunningInit3 {
+						continue
+					}
+					if c.State.Running != nil {
```
As with https://github.com/kubernetes/kubernetes/pull/125935#discussion_r1668380613, it might be better to explicitly check if the long-running init container is start running.

```suggestion
					if c.State.Running != nil &&
						(c.Started == nil || *c.Started == true) {
```

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-containerd

gjkim42 commented:
/test pull-kubernetes-node-kubelet-serial-containerd

gjkim42 commented:
/retest

SergeyKanzhelev reviewed (approved):
/lgtm
/approve
